### ISSUE TITLE
"odo delete -a" with "--app" & "--project" flags deletes the component under cwd/pwd

### ISSUE THREAD

**Issue by dharmit** [openshift/odo#4451]
/kind bug

## What versions of software are you using?

**Operating System:** all

**Output of `odo version`:** `odo v2.0.5 (2180507a4)`

## How did you run odo exactly?
Tried to delete component named `node` using the command `odo delete node --app app --project myproject -af` from the component directory of `node-todo` component.

## Actual behavior
odo deletes the `node-todo` component instead of deleting `node` component. Since it has `-a` flag, it also deletes underlying config files (`.odo` directory and `devfile.yaml`.)

## Expected behavior
odo should delete `node` component as specified on the CLI.

## Steps to reproduce the issue
1. Create and push two components, let's say, `comp1` and `comp2`.
2. From the component directory of `comp2`, execute the component deletion command:
    ```sh
    $ odo delete comp1 --app app --project proj --all
    ```
3. odo deletes `comp2` instead of deleting `comp1`.

/area component

**Comment by dharmit** [comment]
As I mentioned on https://github.com/openshift/odo/issues/4135#issuecomment-783174485, I think we should have a validation that prohibits the user from passing `-a`/`--all` flag when they're trying to delete a component from a directory other than the component's own directory. So, a user can't pass `--all` flag if they have passed `--app` and `--project`. 

AFAIK,  `--app` and `--project` both must be provided to delete a component just by its name.

**Comment by kadel** [comment]
> odo deletes the `node-todo` component instead of deleting `node` component. Since it has `-a` flag, it also deletes underlying config files (`.odo` directory and `devfile.yaml`.)

This is actually correct behavior. From odo points of view, this is is what you are telling it to do.
But this behavior might be confusing for users. 






> AFAIK, `--app` and `--project` both must be provided to delete a component just by its name.

Nope, you can use just one. For example in a directory with `nodejs ` component deployed into `app ` application in `test1 ` project.
If you run  `odo delete --project test2` odo will try to delete component `nodejs ` in `app `application in `test2 `project.

This is how it was originally intentionally designed But I'm not sure if this makes sense anymore. 

We could consider changing it. We might try to make it a little bit simpler. The user would have only two options. Either use odo config files (devfiles) or define the component manually.  By that, I mean that we would also forbid partial specification like `odo delete nodejs`


examples of **invalid** commands:
-  `odo delete nodejs --context ./dir` 
-  `odo delete nodejs --all` 
-  `odo delete  --project myproj --context ./dir` 
-  `odo delete  --project myproj` 
- `odo delete nodejs --app myapp --project myproj --all`


**valid** command examples:
- `odo delete --context ./my-component --all`
- `odo delete --context ./my-component`
- `odo delete`
- `odo delete --all`
- `odo delete nodejs --app myapp --project myproj`

The list of valid commands could be complete in this case. There would be no additional valid `odo delete` commands.

**Comment by valaparthvi** [comment]
Scope of this issue for Sprint 199 -
1. Play with `odo delete` and note observations.

_I think 2 days(including buffer) would be enough for this issue._

**Comment by valaparthvi** [comment]
**tldr;**
1. `odo delete` does not respect `--context` when it comes to deleting `devfile.yaml`.
2. `odo delete` does not respect `--app` or `--project` when it comes to deleting the component, it only takes into account the component present in the cwd.

---
Consider the following setup - 
`tuxDir` with component `tuxComp` belonging to app  `tuxApp` and project `tuxProject`.
`chuckDir` with component `chuckComp` belonging to app  `chuckApp` and project `chuckProject`.

With the above setup and `tuxDir` as the cwd, here are my observations - 
-  `odo delete --context chuckDir --all` deletes the correct component and `.odo`, but deletes `tuxDir`'s `devfile.yaml`
- `odo delete chuckComp --all` asks if the user wants to delete `tuxComp` and using `-f` deletes the `tuxComp`
- `odo delete chuckComp --project chuckProject` asks if the user wants to delete the `tuxComp` and using `-f` deletes the `tuxComp`

The above observations stay true  _regardless of the component's state(not/pushed)_.

The valid commands @kadel suggested in https://github.com/openshift/odo/issues/4451#issuecomment-783431524 would make things much simpler and are intuitive.

IIUC, 
-  `odo delete --context ./my-component --all` - will delete the component, devfile.yaml and .odo in the context dir, _regardless of the component's state(not/pushed)_. 
-  `odo delete --context ./my-component`- will delete the component in the context dir,_regardless of the component's state(not/pushed)_.
-  `odo delete` - will delete the component in the cwd, _regardless of the component's state(not/pushed)_.
-  `odo delete --all`-  - will delete the component, devfile.yaml and .odo in the cwd, _regardless of the component's state(not/pushed)_.
-  `odo delete nodejs --app myapp --project myproj` - will delete the component in app and project regardless of the cwd. _But this might not work in case the component is not pushed._

Note: If it helps, I can put the above scenarios in a test-file, I already have a file half-written.

**Comment by valaparthvi** [comment]
Scope of the issue for Sprint 200 - 
- [ ] 1. Add missing integration test cases
- [ ] 2. Test with `--wait` command and add integration tests if not added.

**Comment by valaparthvi** [comment]
`--wait` command is broken.
```
emptyDir2 odo delete; kubectl get pods       
? Are you sure you want to delete the devfile component: nodejs-emptydir2-uygl? Yes
Gathering information for component nodejs-emptydir2-uygl
 ✓  Checking status for component [11ms]
Deleting component nodejs-emptydir2-uygl
 ✓  Deleting Kubernetes resources for component [7ms]
 ✓  Successfully deleted component
NAME                                    READY   STATUS        RESTARTS   AGE
nodejs-emptydir2-uygl-b97dd94d5-zt4gd   1/1     Terminating   0          19s
```

```
➜  emptyDir2 odo delete --wait; kubectl get pods
? Are you sure you want to delete the devfile component: nodejs-emptydir2-uygl? Yes
Gathering information for component nodejs-emptydir2-uygl
 ✓  Checking status for component [10ms]
Deleting component nodejs-emptydir2-uygl
 ✓  Deleting Kubernetes resources for component [7ms]
 ✓  Successfully deleted component
NAME                                    READY   STATUS        RESTARTS   AGE
nodejs-emptydir2-uygl-b97dd94d5-mp5w6   1/1     Terminating   0          20s
```

This seems to be related to #4593.

**Comment by valaparthvi** [comment]
Scope for sprint 201 - 
- [ ] Merge #4641 
- [ ] Fix [this](https://github.com/openshift/odo/pull/4641/files#diff-8462e58763e4f95e797185b6557e5a8608ec2bf2ce7744c06f94d53f48db4d10R189 ) - deleting the correct devfile.yaml if `--context` flag is passed.

**Comment by valaparthvi** [comment]
Scope for sprint 202 -

- [ ] Merge #4641
- [ ] Make sure there are tests for both valid and invalid commands.
- [ ] Add checks for invalid commands.
